Simplified/AxisService/NYPLAxisDownloadProgress.swift

```diff
@@ -0,0 +1,192 @@
+//  Open eBooks
+//
+//  Created by Raman Singh on 2021-05-27.
+//  Copyright © 2021 NYPL Labs. All rights reserved.
+//
+import Foundation
+
+protocol NYPLAxisDownloadProgressHandling: class {
+  var getCurrentProgress: Double { get }
+  var progressListener: NYPLAxisDownloadProgressListening? { get set }
+  func addFixedWeightTask(with url: URL, weight incomingWeight: Double) throws
+  func addFlexibleWeightTask(with url: URL) throws
+  func didFinishTask(with url: URL)
+}
+
+class NYPLAxisDownloadProgress: NYPLAxisDownloadProgressHandling {
+
+  enum NYPLAxisDownloadProgressError: Error {
+    /// Total weight of fixed items matches or exceeds 1.0
+    case illegalWeightDistribution
+    /// Process already readched 100 %
+    case processAlreadyCompleted
+    /// Item with given url already added
+    case itemAlreadyExists
+  }
+
+  private let allTasks: ThreadSafeValueContainer<[DownloadTask]>
+  private let currentProgress: ThreadSafeValueContainer<Double>
+  weak var progressListener: NYPLAxisDownloadProgressListening?
+
+  var getCurrentProgress: Double {
+    return currentProgress.value ?? 0.0
+  }
+
+  init(queue: DispatchQueue = DispatchQueue(label: "NYPLAxisDownloadProgressQueue")) {
+    self.allTasks = ThreadSafeValueContainer(value: [], queue: queue)
+    self.currentProgress = ThreadSafeValueContainer(value: 0.0, queue: queue)
+  }
+
+  /// Adds a fixed weight download task with given weight.
+  ///
+  /// - Note: Adds the task as flexible weight if the weight provided is less than or equals 0.0, greater
+  /// than 1.0, or if after adding the task the combined weight of all the fixed weight tasks becomes more
+  /// than 1.0.
+  ///
+  /// - Throws: Throws an exception in the following scenarios:
+  /// 1. The item has already been added.
+  /// 2. The download progress has already reached 100%.
+  /// 3. The total weight of fixed weight items is 1.0
+  func addFixedWeightTask(with url: URL, weight incomingWeight: Double) throws {
+    // Added wrong weight amount. Adding it as a flexible weight item instead
+    if (incomingWeight <= 0.0 || incomingWeight > 1.0) {
+      try addFlexibleWeightTask(with: url)
+      return
+    }
+
+    // If adding the item will result in weight higher than 1.0, we add it as a
+    // flexible weight item
+    let currentFixedWeight = combinedWeightOfAllFixedWeightTasks()
+    guard currentFixedWeight + incomingWeight <= 1.0 else {
+      try addFlexibleWeightTask(with: url)
+      return
+    }
+
+    try addTask(
+      DownloadTask(url: url, weight: incomingWeight, isFixedWeight: true))
+  }
+
+  /// Adds a flexible weight download task.
+  ///
+  /// - Throws: Throws an exception in the following scenarios:
+  /// 1. The item has already been added.
+  /// 2. The download progress has already reached 100%.
+  /// 3. The total weight of fixed weight items is 1.0
+  func addFlexibleWeightTask(with url: URL) throws {
+    try addTask(DownloadTask(url: url, weight: 1.0, isFixedWeight: false))
```

Comment:
When we add a task with flexible weight, before it's added to the tasks array, the weight of all the unfinished flexible tasks is adjusted so that 
`totalOfFixedWeightTasks + totalOfCompletedFlexibleWeightTasks + totalOfUnfinishedFlexibleWeightTasks = 1.0.`

Therefore, regardless of what weight we assign to a task with flexible weight, its weight is calculated to accommodate the above condition.